### PR TITLE
Initial PKCE implementation

### PR DESCRIPTION
**Pull request by ulrikstrid**

### CODE REVIEW

File: oidc/pkce.ml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+let alphabet = Base64.make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+
+let octets = 96 (* 4 * (96/3) = 128 *)
+
+let base64_encode_cstruct cstruct =
+  Base64.encode_string ~alphabet ~pad:false @@ Cstruct.to_string cstruct
+
+type verifier = string
+
+(* https://www.rfc-editor.org/rfc/rfc7636#section-4.2 *)
```
Comment (anmonteiro) [inline oidc/pkce.ml]: I tend to like submodules, I think this could be `Pkce.Challenge.t`

File: oidc/pkce.ml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  S256 s256_challenge_string
+
+let challenge_of_string ~transformation challenge =
+  match transformation with
+  | `S256 -> S256 challenge
+  | `Plain -> Plain challenge
+
+(* https://www.rfc-editor.org/rfc/rfc7636#section-4.3 *)
+let challenge_to_code_challange_and_method challenge =
```
Comment (anmonteiro) [inline oidc/pkce.ml]:
```suggestion
let challenge_to_code_challenge_and_method challenge =
```

File: oidc/pkce.mli [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+val verifier_of_string : string -> verifier
+
+(** Will always be a S256 challenge
+  https://www.rfc-editor.org/rfc/rfc7636#section-4.2 *)
+val make_code_challenge : verifier -> challenge 
+
+val challenge_of_string : transformation:challenge_transformation -> string -> challenge
+
+(** https://www.rfc-editor.org/rfc/rfc7636#section-4.3 *)
+val challenge_to_code_challange_and_method : challenge -> string * string
```
Comment (anmonteiro) [inline oidc/pkce.mli]:
```suggestion
val challenge_to_code_challenge_and_method : challenge -> string * string
```

File: oidc/pkce.ml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    can also contain "." and "~" but we already have 64 characters  
+*)
+let alphabet = Base64.make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+
+let octets = 96 (* 4 * (96/3) = 128 *)
+
+let base64_encode_cstruct cstruct =
+  Base64.encode_string ~alphabet ~pad:false @@ Cstruct.to_string cstruct
+
+type verifier = string
```
Comment (anmonteiro) [inline oidc/pkce.ml]: probably a `module Verifier` too?

File: oidc/pkce.ml [hunk 5]
```diff
@@ -0,0 +1,61 @@
+(**
+ * Copyright 2022 Ulrik Strid. All rights reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file.
+ *)
+
+(* https://www.rfc-editor.org/rfc/rfc3986#section-2.3
+    can also contain "." and "~" but we already have 64 characters  
+*)
+let alphabet = Base64.make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
```
Comment (anmonteiro) [inline oidc/pkce.ml]: isn't this `Base64.uri_safe_alphabet`?

Comment (ulrikstrid) [reply oidc/pkce.ml]: I think so, but we might want to change it in the future, but I guess I could just do an alias now?

Comment (anmonteiro) [reply oidc/pkce.ml]: Probably good to be explicit. let's keep it as is

**Pull request merged**